### PR TITLE
feat(nms): migrate /nms/app/view/alarms to typescript

### LINKED ISSUE #12815: Migrate /nms/views/alarms to TypesScript
Migrate everything in `/nms/views/alarms` and everything that is imported from `/nms/views/alarms` to TypeScript.

This includes the following files:

- [x] ../../state/AlarmsApiUtil.js
- [x] ../../state/SyncAlerts.js
- [x] AlarmsDashboard.js
- [x] components/AlarmAPIType.js
- [x] components/AlarmContext.js
- [x] components/Alarms.js
- [x] components/AlarmsApi.js
- [x] components/AlertRules.js
- [x] components/ClipboardLink.js
- [x] components/alertmanager/AlertDetails/AlertDetailsPane.js
- [x] components/alertmanager/AlertDetails/MetricAlertViewer.js
- [x] components/alertmanager/AlertDetails/\_\_tests\_\_/AlertDetailsPane-test.js
- [x] components/alertmanager/FiringAlerts.js
- [x] components/alertmanager/Receivers/AddEditReceiver.js
- [x] components/alertmanager/Receivers/ConfigEditor.js
- [x] components/alertmanager/Receivers/EmailConfigEditor.js
- [x] components/alertmanager/Receivers/GlobalConfig.js
- [x] components/alertmanager/Receivers/PagerDutyConfigEditor.js
- [x] components/alertmanager/Receivers/PushoverConfigEditor.js
- [x] components/alertmanager/Receivers/Receivers.js
- [x] components/alertmanager/Receivers/SelectReceiver.js
- [x] components/alertmanager/Receivers/SlackConfigEditor.js
- [x] components/alertmanager/Receivers/WebhookConfigEditor.js
- [x] components/alertmanager/Receivers/\_\_tests\_\_/AddEditReceiver-test.js
- [x] components/alertmanager/Receivers/\_\_tests\_\_/GlobalConfig-test.js
- [x] components/alertmanager/Receivers/\_\_tests\_\_/Receivers-test.js
- [x] components/alertmanager/Routes.js
- [x] components/alertmanager/Suppressions.js
- [x] components/alertmanager/\_\_tests\_\_/FiringAlerts-test.js
- [x] components/common/Editor.js
- [x] components/hooks.js
- [x] components/prometheus/PromQL.js
- [x] components/prometheus/PromQLParser.js
- [x] components/prometheus/PromQLTokenizer.js
- [x] components/prometheus/PromQLTypes.js
- [x] components/prometheus/\_\_generated\_\_/PromQLGrammar.js (not flow and generated -> ignored)
- [x] components/prometheus/\_\_tests\_\_/PromQL-test.js
- [x] components/prometheus/\_\_tests\_\_/PromQLParser-test.js
- [x] components/rules/AddEditRule.js
- [x] components/rules/LabelsEditor.js
- [x] components/rules/PrometheusEditor/PrometheusEditor.js
- [x] components/rules/PrometheusEditor/ToggleableExpressionEditor.js
- [x] components/rules/PrometheusEditor/\_\_tests\_\_/PrometheusEditor-test.js
- [x] components/rules/PrometheusEditor/\_\_tests\_\_/ToggleableExpressionEditor-test.js
- [x] components/rules/PrometheusEditor/getRuleInterface.js
- [x] components/rules/PrometheusEditor/index.js
- [x] components/rules/RuleContext.js
- [x] components/rules/RuleEditorBase.js
- [x] components/rules/RuleInterface.js
- [x] components/rules/SelectRuleType.js
- [x] components/rules/\_\_tests\_\_/AddEditRule-test.js
- [x] components/rules/\_\_tests\_\_/LabelsEditor-test.js
- [x] components/severity/Severity.js
- [x] components/severity/SeverityIndicator.js
- [x] components/table/SimpleTable.js
- [x] components/table/TableActionDialog.js
- [x] components/table/TableAddButton.js
- [x] components/table/\_\_tests\_\_/SimpleTable-test.js
- [x] hooks/useForm.js
- [x] legacy/Alarms.js
- [x] legacy/\_\_tests\_\_/Alarms-test.js
- [x] test/testData.js
- [x] test/testHelpers.js

Image

### PR DESCRIPTION
**Pull request by thmsschmitt**
## Summary

- Requires:
  - https://github.com/magma/magma/pull/12823
- This is **part 2** of the actual NMS migration to TypeScript. 
- Fixes https://github.com/magma/magma/issues/12815

## Test Plan

- `yarn flow`
- `yarn tsc`
- `yarn run eslint . --quiet`
- `yarn test`

### CODE REVIEW

File: nms/app/components/context/AppContext.tsx [hunk 1]
```diff
@@ -9,63 +9,66 @@
...
-  ssoSelectedType: SSOSelectedType,
-  hasAccountSettings: boolean,
+  csrfToken: string | null | undefined;
+  version: string | null | undefined;
+  networkIds: Array<string>;
+  user: User;
+  showExpandButton: () => void;
+  hideExpandButton: () => void;
+  isOrganizations: boolean;
+  isFeatureEnabled: (arg0: FeatureID) => boolean;
```
Comment (thmsschmitt) [inline nms/app/components/context/AppContext.tsx]: `arg0` -> `featureId`

File: nms/app/components/context/AppContext.tsx [hunk 2]
```diff
@@ -9,63 +9,66 @@
...
+    appData: EmbeddedData;
+  } = window.CONFIG;
   const {appData} = config;
   const value = {
     ...appContextDefaults,
     ...appData,
     hasAccountSettings: !appData.ssoEnabled,
-    isOrganizations: !!props.isOrganizations, // is organizations management aka. the host site
+    isOrganizations: !!props.isOrganizations,
+    // is organizations management aka. the host site
```
Comment (thmsschmitt) [inline nms/app/components/context/AppContext.tsx:70]: Should be above `isOrganizations`

File: nms/app/components/ActionTable.tsx [hunk 3]
```diff
@@ -30,145 +26,122 @@ import FilterList from '@material-ui/icons/FilterList';
...
-  options: ActionTableOptions,
-  detailPanel?: Array<{}>,
-  onSelectionChange?: (Array<T>) => void,
+export type ActionTableProps<T extends object> = {
+  titleIcon?: typeof SvgIcon;
+  tableRef?: MaterialTableProps<T>['tableRef'];
+  editable?: MaterialTableProps<T>['editable'];
+  localization?: MaterialTableProps<T>['localization'];
+  title?: string;
+  handleCurrRow?: (arg0: T) => void;
```
Comment (thmsschmitt) [inline nms/app/components/ActionTable.tsx]: `arg0` -> `currentRow`

File: nms/app/components/ActionTable.tsx [hunk 4]
```diff
@@ -30,145 +26,122 @@ import FilterList from '@material-ui/icons/FilterList';
...
 type SelectProps = {
-  content: Array<string>,
-  defaultValue?: string,
-  value: string,
-  onChange: string => void,
-  testId?: string,
+  content: Array<string>;
+  defaultValue?: string;
+  value: string;
+  onChange: (arg0: string) => void;
```
Comment (thmsschmitt) [inline nms/app/components/ActionTable.tsx]: `arg0` -> `value`

File: nms/app/views/subscriber/SubscriberUtils.ts [hunk 5]
```diff
@@ -91,24 +88,21 @@ const SUBSCRIBER_ACTION_TYPE = Object.freeze({
...
+  authOpc: string;
+  state: 'INACTIVE' | 'ACTIVE';
+  forbiddenNetworkTypes: Array<SubscriberForbiddenNetworkTypesEnum>;
+  dataPlan: string;
+  apns: Array<string>;
+  policies?: Array<string>;
 };
-
-type SubscriberError = $Values<typeof SUBSCRIBER_ADD_ERRORS>;
+type SubscriberError = typeof SUBSCRIBER_ADD_ERRORS[keyof typeof SUBSCRIBER_ADD_ERRORS];
```
Comment (thmsschmitt) [inline nms/app/views/subscriber/SubscriberUtils.ts]: We can reuse this type below

File: nms/app/theme/design-system/Text.tsx [hunk 6]
```diff
@@ -9,17 +9,14 @@
...
- * @format
  */
 
 import * as React from 'react';
 import classNames from 'classnames';
 import {makeStyles} from '@material-ui/styles';
 import {typography} from '../default';
 
-const styles = {
+export const typographyStyles = makeStyles(() => ({
```
Comment (thmsschmitt) [inline nms/app/theme/design-system/Text.tsx]: Can be shortened to `makeStyles({ ...`

File: nms/app/theme/design-system/Text.tsx [hunk 7]
```diff
@@ -71,10 +68,10 @@ type Props = {
     | 'body2'
     | 'body3'
     | 'caption'
-    | 'overline',
-  className?: string,
-  useEllipsis?: ?boolean,
-  weight?: 'inherit' | 'light' | 'regular' | 'medium' | 'bold',
+    | 'overline';
+  className?: string;
+  useEllipsis?: boolean | null | undefined;
```
Comment (thmsschmitt) [inline nms/app/theme/design-system/Text.tsx]: Nit pick having both `?` and `| undefined` is somewhat redundant in this context

File: nms/app/views/network/FEGNetworkInfo.tsx [hunk 8]
```diff
@@ -9,40 +9,35 @@
...
-
-// $FlowFixMe[cannot-resolve-module] for TypeScript migration
 import DataGrid from '../../components/DataGrid';
 import React from 'react';
+import type {DataRows} from '../../components/DataGrid';
+import type {FegNetwork} from '../../../generated-ts';
 
 type Props = {
-  fegNetwork: $Shape<feg_network>,
+  // TODO: Is a partial really needed here? Also fix the implications for NetworkInfo.
```
Comment (thmsschmitt) [inline nms/app/views/network/FEGNetworkInfo.tsx]: `TODO[TS-migration] ...`

File: nms/app/components/FEGServicingAccessGatewayKPIs.tsx [hunk 9]
```diff
@@ -9,55 +9,46 @@
...
+  federationNetworkId: NetworkId,
   enqueueSnackbar?: (
     msg: string,
-    cfg: EnqueueSnackbarOptions,
-  ) => ?(string | number),
-): Promise<Array<feg_lte_network>> {
-  const servicedAccessNetworks = [];
-  const fegLteNetworkIdList = await MagmaV1API.getFegLte();
+    cfg: OptionsObject,
+  ) => (string | number) | null | undefined,
```
Comment (thmsschmitt) [inline nms/app/components/FEGServicingAccessGatewayKPIs.tsx]: the brackets around `string | number` are not needed

File: nms/app/views/network/FEGServicingAccessGatewayTable.tsx [hunk 10]
```diff
@@ -57,56 +44,59 @@ type ServicingAccessGatewayRowType = {
...
  */
 async function getServicedAccessGatewaysInfo(
-  servicedAccessNetworks: Array<feg_lte_network>,
+  servicedAccessNetworks: Array<FegLteNetwork>,
   enqueueSnackbar: (
     msg: string,
-    cfg: EnqueueSnackbarOptions,
-  ) => ?(string | number),
+    cfg: OptionsObject,
+  ) => (string | number) | null | undefined,
```
Comment (thmsschmitt) [inline nms/app/views/network/FEGServicingAccessGatewayTable.tsx]: bracket around `string | number` are not needed

File: nms/app/views/network/__tests__/FEGServicingAccessGatewayTableTest.tsx [hunk 11]
```diff
@@ -105,12 +100,11 @@ const mockGw1: lte_gateway = {
   checked_in_recently: false,
 };
 
-jest.mock('axios');
-jest.mock('../../../../generated/MagmaAPIBindings.js');
+//jest.mock('axios');
```
Comment (thmsschmitt) [inline nms/app/views/network/__tests__/FEGServicingAccessGatewayTableTest.tsx]: remove

File: nms/app/views/network/__tests__/FEGServicingAccessGatewayTableTest.tsx [hunk 12]
```diff
@@ -190,31 +190,47 @@ describe('<ServicingAccessGatewaysInfo />', () => {
       </MemoryRouter>
     );
   };
+
   it('renders serviced access gateway table correctly', async () => {
     const {getAllByRole} = render(<Wrapper />);
     await wait();
     //first get list of feg_lte networks
-    expect(MagmaAPIBindings.getFegLte).toHaveBeenCalledTimes(1);
+    /* eslint-disable @typescript-eslint/unbound-method */
```
Comment (thmsschmitt) [inline nms/app/views/network/__tests__/FEGServicingAccessGatewayTableTest.tsx]: not needed

File: nms/app/views/alarms/components/alertmanager/Suppressions.tsx [hunk 13]
```diff
@@ -45,31 +39,28 @@ const useStyles = makeStyles(theme => ({
...
   const [showDialog, setShowDialog] = useState(false);
-  const [lastRefreshTime, _setLastRefreshTime] = useState<string>(
-    new Date().toLocaleString(),
-  );
-  const [_isAddEditAlert, _setIsAddEditAlert] = useState<boolean>(false);
+  const [lastRefreshTime] = useState<string>(new Date().toLocaleString());
   const classes = useStyles();
   const snackbars = useSnackbars();
   const networkId = useNetworkId();
+  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
```
Comment (thmsschmitt) [inline nms/app/views/alarms/components/alertmanager/Suppressions.tsx]: Do we really need this?

File: nms/app/views/alarms/components/alertmanager/AlertDetails/AlertDetailsPane.tsx [hunk 14]
```diff
@@ -151,6 +175,7 @@ function AlertTroubleshootingLink({alertName}: {alertName: string}) {
   const classes = useStyles();
   const snackbars = useSnackbars();
   const {apiUtil} = useAlarmContext();
+  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
```
Comment (thmsschmitt) [inline nms/app/views/alarms/components/alertmanager/AlertDetails/AlertDetailsPane.tsx]: Looks like this is also not needed?

File: nms/app/views/alarms/test/testHelpers.tsx [hunk 15]
```diff
@@ -9,42 +9,33 @@
...
+export function mockApiUtil(merge?: Partial<ApiUtil>): ApiUtil {
   /**
    * I don't understand how to properly type these mocks so using any for now.
    * The consuming code is all strongly typed, this shouldn't be much of an issue.
    */
   // eslint-disable-next-line flowtype/no-weak-types
   const useAlarmsApi = jest.fn<any, any>(
     <TParams, TResponse>(
-      func: TParams => Promise<TResponse>,
+      func: (arg0: TParams) => Promise<TResponse>,
```
Comment (thmsschmitt) [inline nms/app/views/alarms/test/testHelpers.tsx]: `arg0` -> `params`

**Review by thmsschmitt — COMMENTED** [review]
The changes from @Neudrino look good to me. I left a few minor review comments.

File: nms/app/components/context/AppContext.tsx (hunk 1 above)
Comment (Neudrino) [reply nms/app/components/context/AppContext.tsx]: Already done with https://github.com/magma/magma/pull/12823#discussion_r885516547.

File: nms/app/components/context/AppContext.tsx (hunk 2 above)
Comment (Neudrino) [reply nms/app/components/context/AppContext.tsx:70]: Already done with https://github.com/magma/magma/pull/12823#discussion_r885517283

File: nms/app/components/ActionTable.tsx (hunk 4 above)
Comment (Neudrino) [reply nms/app/components/ActionTable.tsx]: Already done with https://github.com/magma/magma/pull/12823#discussion_r885526846.

File: nms/app/views/subscriber/SubscriberUtils.ts (hunk 5 above)
Comment (Neudrino) [reply nms/app/views/subscriber/SubscriberUtils.ts]: Already done with https://github.com/magma/magma/pull/12823#discussion_r885529728.

File: nms/app/theme/design-system/Text.tsx (hunk 6 above)
Comment (Neudrino) [reply nms/app/theme/design-system/Text.tsx]: Already done with https://github.com/magma/magma/pull/12823#discussion_r885532691.

File: nms/app/views/network/FEGNetworkInfo.tsx (hunk 8 above)
Comment (Neudrino) [reply nms/app/views/network/FEGNetworkInfo.tsx]: Already done with https://github.com/magma/magma/pull/12823#discussion_r885537568.

File: nms/app/components/FEGServicingAccessGatewayKPIs.tsx (hunk 9 above)
Comment (Neudrino) [reply nms/app/components/FEGServicingAccessGatewayKPIs.tsx]: Did IntelliJ full text search and replace.

File: nms/app/views/network/__tests__/FEGServicingAccessGatewayTableTest.tsx (hunk 11 above)
Comment (Neudrino) [reply nms/app/views/network/__tests__/FEGServicingAccessGatewayTableTest.tsx]: Already done with https://github.com/magma/magma/pull/12823#discussion_r885542789.

File: nms/app/views/alarms/components/alertmanager/Suppressions.tsx (hunk 13 above)
Comment (Neudrino) [reply nms/app/views/alarms/components/alertmanager/Suppressions.tsx]: Got removed elsewhere.

File: nms/app/views/alarms/components/ForwardRef.js [hunk 16]
```diff
@@ -1,58 +0,0 @@
-/**
```
Comment (Neudrino) [inline nms/app/views/alarms/components/ForwardRef.js:1]: Deleted without replacement?

File: nms/app/state/AlarmsApiUtil.ts [hunk 17]
```diff
@@ -0,0 +1,162 @@
+/*
```
Comment (Neudrino) [inline nms/app/state/AlarmsApiUtil.ts:1]: _The diff does not display well for this file in Github. I just trust it is good._

**Review by Neudrino — APPROVED** [review]
Reviewed commits by @thmsschmitt starting from [refactor(nms): Migrate AlarmAPIType to TypeScript](https://github.com/magma/magma/pull/12886/commits/15b65294ba3fb66746b2ce211d54931a5aa6b2cb).

Only view minor comments.

File: nms/app/views/alarms/components/ForwardRef.js (hunk 16 above)
Comment (thmsschmitt) [reply nms/app/views/alarms/components/ForwardRef.js:1]: Yes it turned to be unused

**Comment by Neudrino** [comment]
Closed in favour of the big **NMS migration PR**:
- https://github.com/magma/magma/pull/13098

**Pull request closed without merging**